gear/gear/database.py

```diff
@@ -16,45 +24,148 @@
         cursorclass=aiomysql.cursors.DictCursor, autocommit=autocommit)
 
 
+class TransactionAsyncContextManager:
+    def __init__(self, db_pool, read_only):
+        self.db_pool = db_pool
+        self.read_only = read_only
+        self.tx = None
+
+    async def __aenter__(self):
+        tx = Transaction()
+        await tx.async_init(self.db_pool, self.read_only)
+        self.tx = tx
+        return tx
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.tx._aexit(exc_type, exc_val, exc_tb)
+        self.tx = None
+
+
+class Transaction:
+    def __init__(self):
+        self.conn_context_manager = None
+        self.conn = None
+
+    async def async_init(self, db_pool, read_only):
+        self.conn_context_manager = db_pool.acquire()
+        self.conn = await aenter(self.conn_context_manager)
+        async with self.conn.cursor() as cursor:
+            if read_only:
+                await cursor.execute('START TRANSACTION READ ONLY;')
+            else:
+                await cursor.execute('START TRANSACTION;')
+
+    async def _aexit(self, exc_type, exc_val, exc_tb):
+        if not self.conn:
+            return
+
+        if exc_type:
+            await self.conn.rollback()
+        else:
+            await self.conn.commit()
+        self.conn = None
+
+        await aexit(self.conn_context_manager, exc_type, exc_val, exc_tb)
+        self.conn_context_manager = None
+
+    async def commit(self):
+        assert self.conn
+        self.conn.commit()
+        self.conn = None
+
+        await aexit(self.conn_context_manager)
+        self.conn_context_manager = None
+
+    async def rollback(self):
```

Comment:
this is a bit of a weird method, no? After a rollback, there's nothing you could possibly do with this transaction, so why still be inside the transaction's context? I'd think there would be no `commit` and `rollback` raised a special exception that was caught by `exit` which rolled back but did not propagate the exception.

Reply:
You don't have to use a transaction in a context (e.g. you want to store a long-lived tx in a class member).